cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,380 @@
+        Set<Broker> allBrokersInDataStore = brokersByBrokerSet.values()
+                .stream()
+                .flatMap(brokers -> brokers.stream())
+                .collect(Collectors.toSet());
+
+        if (!allBrokersInDataStore.equals(clusterModel.brokers())) {
+            throw new OptimizationFailureException(String.format("[%s] All Brokers from data store %s do not match brokers in cluster model %s.",
+                                                                 name(), allBrokersInDataStore, clusterModel.brokers()));
+        }
+
+        for (Map.Entry<String, List<Partition>> partitionsByTopic : clusterModel.getPartitionsByTopic()
+                .entrySet()) {
+            if (!excludedTopics.contains(partitionsByTopic.getKey())) {
+                List<Partition> partitions = partitionsByTopic.getValue();
+                Set<Broker> allBrokersForTopic = partitions.stream()
+                        .map(partition -> partition.partitionBrokers())
+                        .flatMap(brokers -> brokers.stream())
+                        .collect(Collectors.toSet());
+                // Checks if a partition's brokers do not all live in a single brokerSet
+                if (!brokersByBrokerSet.values()
+                        .stream()
+                        .anyMatch(brokerSetBrokers -> brokerSetBrokers.containsAll(allBrokersForTopic))) {
+                    throw new OptimizationFailureException(String.format("[%s] Topic %s is not brokerSet-aware. brokers (%s).",
+                                                                         name(), partitionsByTopic.getKey(), allBrokersForTopic));
+                }
+            }
+        }
+    }
+
+    /**
+     * brokerSet-awareness violations can be resolved with replica movements.
+     *
+     * @param broker              Broker to be balanced.
+     * @param clusterModel        The state of the cluster.
+     * @param optimizedGoals      Optimized goals.
+     * @param optimizationOptions Options to take into account during optimization.
+     */
+    @Override
+    protected void rebalanceForBroker(Broker broker,
+                                      ClusterModel clusterModel,
+                                      Set<Goal> optimizedGoals,
+                                      OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+        LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+        Map<String, Set<Broker>> brokersByBrokerSet = _brokerSetDataHelper.getBrokersByBrokerSet();
+        Map<Broker, String> brokerSetByBroker = _brokerSetDataHelper.getBrokerSetByBroker();
+
+        String currentBrokerSetId = brokerSetByBroker.get(broker);
+        if (currentBrokerSetId == null) {
+            throw new OptimizationFailureException(String.format("[%s] Cannot find broker set for broker id %s", name(), broker.id()));
+        }
```

Comment:
Can we ensure that this can't happen by construction of `brokerSetByBroker` and skip sanity check here?